v2/scheduler/internal/lib/queue/amqp/reader.go

```diff
@@ -0,0 +1,204 @@
+		},
+		amqpClientMu: &sync.Mutex{},
+	}
+	q.connectFn = q.connect
+	return q
+}
+
+// connect connects (or reconnects) to the underlying AMQP-based messaging
+// server. This function is NOT concurrency safe and callers should take
+// measures to ensure they are the exclusive caller of this function.
+func (q *readerFactory) connect() error {
+	return retries.ManageRetries(
+		context.Background(),
+		"connect",
+		10,
+		10*time.Second,
+		func() (bool, error) {
+			if q.amqpClient != nil {
+				q.amqpClient.Close()
+			}
+			var err error
+			if q.amqpClient, err = myamqp.Dial(q.address, q.dialOpts...); err != nil {
+				return true, errors.Wrap(err, "error dialing endpoint")
+			}
+			return false, nil
+		},
+	)
+}
+
+func (q *readerFactory) NewReader(queueName string) (queue.Reader, error) {
+	// This entire function is a critical section of code so that we don't
+	// possibly have multiple callers looking for a new Reader opening multiple
+	// underlying connections to the messaging server.
+	q.amqpClientMu.Lock()
+	defer q.amqpClientMu.Unlock()
+
+	if q.amqpClient == nil {
+		if err := q.connectFn(); err != nil {
+			return nil, err
+		}
+	}
+
+	linkOpts := []amqp.LinkOption{
+		amqp.LinkSourceAddress(queueName),
+		// Link credit is 1 because we're a "slow" consumer. We do not want messages
+		// piling up in a client-side buffer, knowing that it could be some time
+		// before we can process them.
+		amqp.LinkCredit(1),
+	}
+
+	// Every Writer will get its own Session and Receiver
+	var amqpSession myamqp.Session
+	var amqpReceiver myamqp.Receiver
+	var err error
+	for {
+		// If we've been through the loop before, try cleaning up the session and/or
+		// receiver that we never ended up using.
+		if amqpReceiver != nil {
+			amqpReceiver.Close(context.TODO()) // nolint: errcheck
+		}
+		if amqpSession != nil {
+			amqpSession.Close(context.TODO()) // nolint: errcheck
+		}
+
+		if amqpSession, err = q.amqpClient.NewSession(); err != nil {
+			// Assume this happened because the existing connection is no good. Try
+			// to reconnect.
+			if err = q.connectFn(); err != nil {
+				// The connection function handles its own retries. If we got an error
+				// here, it's pretty serious. Bail.
+				return nil, err
+			}
+			// We're reconnected now, so loop around to try getting a session again.
+			continue
+		}
+		if amqpReceiver, err = amqpSession.NewReceiver(linkOpts...); err != nil {
+			// Assume this happened because the existing connection is no good.
+			// Just loop around now because we not only need a new connection, but
+			// also a new session.
+			continue
+		}
+		break
+	}
+
+	return &reader{
+		queueName:    queueName,
+		amqpSession:  amqpSession,
+		amqpReceiver: amqpReceiver,
+	}, nil
+}
+
+func (q *readerFactory) Close(context.Context) error {
+	if err := q.amqpClient.Close(); err != nil {
+		return errors.Wrapf(err, "error closing AMQP client")
+	}
+	log.Println("DEBUG: closed AMQP-based queue reader factory")
```

Comment:
Is this statement intended to remain?  Will we eventually want to conditionally print based on a TBD log level?

Reply:
Nice catch. I intended to remove that. I think introducing a leveled logger would be a great idea. Our obligations are v2-alpha by end of the year and we have _plenty_ of time after that to focus on stability and other non-functional requirements, so if you're ok with it, let's defer leveled logging until then.

I will, however, remove this line that I'd intended to remove prior to merging.